### PR TITLE
[DRAFT] Autosave by default

### PR DESCRIPTION
**Pull request by shihan42**
These are the first steps to solve #22
Right now, I'd like to get some feedback on two possible pitfalls/problems. That's why this is a draft PR.

The **first problem** arises in regard to proper async methods:
All the methods that are responsible for saving the project are converted to support async I/O, but we can't have that yet. The whole UI framework doesn't use the async/await pattern. This pattern commonly needs "async all the way", starting at the UI events and ending in the method that for example use async file I/O.

I'm not sure if it is feasible or even a good idea at this point to rewrite the UI system. That could be a lot of work with possibly low benefits. But on the other hand, saving even big production sheets never exceeded a handful of milliseconds on my machine. So we could just ignore this (as I do in `Project.QueueAutosave()`). But tests with big real life examples would be nice.

The **second problem** is the need to manually include the call to autosave at every occasion where we want the autosave to trigger (if it is enabled).

At first I tried to attach the autosave to the undo system. But that didn't really make sense, as the undo system registers the status *before* a change. With a typical call such as `project.RecordUndo().pages.Add(page);` the status of the displayed pages is only changed *after* the undo.

Shenanigans with deferred execution after a call to `RecordUndo()` are more a hack than an actual solution and introduce synchronization issues over threads.

Any ideas how we should proceed?

### CODE REVIEW

**Comment by shihan42** [comment]
No comments? No ideas? No opinions?

**Comment by have-fun-was-taken** [comment]
Since @veger and @SWeini are busy, I'll share what I know.

> I'm not sure if it is feasible or even a good idea at this point to rewrite the UI system.

It would take a ton of effort to rewrite the whole UI, so we should consider thoroughly if we want to do it. The chances that we  want to do it are very low.

> we could just ignore [the async problem]

I'm okay with this. If the current implementation didn't bother with that, and it works, then it will likely continue to work. Although that's a bit risky, true.

Regarding to where attach the save-trigger, I don't know. It needs futher investigation.

**Comment by SWeini** [comment]
1) Just don't use async/await
Yes, there is an async API for file access, and there are certainly cases when it will help keeping the UI responsive (large files, accessing network drives, spin-up of idle hard disks), but none of that are relevant for YAFC autosave in my opinion.

Introducing async in UI is not an easy task. It's complex to keep the UI responsive while still maintaining integrity of the data and preventing the user from making conflicting changes. the very low-level UI programming done all over the place will probably make this even more complex than it would be otherwise. Doing so is equivalent to a full rewrite. Definitely don't recommend doing anytime soon.

2) There are a few possible solutions for auto-saving, all of them have their pros and cons.

a) Add an explicit method call at every place that modifies the data. This shouldn't be too many places. Search for usages of the undo system and you'll have all of them. This keeps undo/auto-save two separate features. However, in future changes bugs are more likely to happen if doing changes to one system and forgetting to keep the other in sync.

b) Change the API of the undo system to something more similar to an transaction system. The classes UndoSnapshot/UndoBatch are a good start, just needs a different interface to use.
```
var transaction = model.StartTransaction();
model.MakeSomeChanges(foo, bar);
transaction.Commit();
```

Then it's extremely easy to add the auto-save in the commit step.

c) As hinted by you, trigger the auto-save by the undo system, but defer it. There is `UI.DispatchInMainThread`, which should do the trick for thread synchronization.

d) Auto-save based on a timer, instead of after each change. For large files it might be the better solution performance-wise. Especially if the user is just undoing/redoing a lot of changes quickly, an immediate auto-save is a waste of performance, but required to meet user expectations.

3) Additional Notes

Personally, I don't care. I won't use the auto-save feature because I regularly just play around and explicitly don't want to save my changes. So any solution that doesn't massively deteriorate the code quality will get my approval.

Make sure to have a proper "backup" system for autosave. Don't overwrite the original file, instead create separate autosave files. At least write to a new file and perform a rename afterwards. Otherwise it's too easy to lose the data.

**Comment by SWeini** [comment]
FYI: I can give you my 3.5 MB pyanodon yafc file for testing performance. I'm still on some older version, so getting it to run flawlessly needs some more detailed instructions.

**Comment by veger** [comment]
Holidays here, so the kids take quite a bunch of my time...

I am not a C# expert (I only did a pet project here and there with it), so these kind of low-level, language specific things are not for me to comment on. I do have some generic advice/feedback:

Are save times even a bottleneck? When I save my py yafc file I don't even notice it (must admit it is an order of magnitude smaller than SWeini's one). If it is not an issue: don't try to fix it by throwing complexity to the (non existing) problem.

> 2 There are a few possible solutions for auto-saving, all of them have their pros and cons.

e) Use some event/trigger system (again not a C# expert but I would assume there is some build-in off-the-self solution) with 'pre-change' and 'post-change' events. Then other parts of the code can benefit from this: undo system, recalculation triggers, auto-save, etc.

Similar as for 2a, use the places of the undo system to add the triggers to fire the events. Maybe accept some lambda input so it is easier to fire the pre and post triggers in a single go.

> Personally, I don't care. I won't use the auto-save feature because I regularly just play around and explicitly don't want to save my changes.

Same, so lets agree to make this optional as it would mess up my way of trying things with YAFC: for pyanodon I need to try a lot before I find something that makes sense, and often it is revamping half of my recipes to make it work.

**Comment by shihan42** [comment]
@SWeini I'd love to make some tests with your large project.

@veger
I'm not sure if there is such an event/trigger system in C#. Can you provide an example in other languages so that I can do some research?

In general, all the points of you both are something to think about. Give me some time for that ;-)

**Comment by veger** [comment]
I remember seeing such a system in Unity (also C#), but it looks like it is part of their library: [UnityEvent](https://docs.unity3d.com/ScriptReference/Events.UnityEvent.html)

There is 'some event', listeners can be added and whenever something triggers, a function (`Invoke()` in this case) is called, causing all subscribed listeners to be called.

I know I saw this scheme in other languages, but I do not have a more active memory of using them. But this is what I (quickly) found:
* https://www.geeksforgeeks.org/node-js-events/
* https://stackoverflow.com/questions/605169/how-to-do-events-in-ruby

Basically it just keeps a list of interested parties (listeners) and call them all in a fire/invoke/emit function. This way the system is (more) decoupled as the 'fire logic' does not need to know what needs to get triggered.

**Comment by shihan42** [comment]
Ok, the event handling won't solve the problem either, since it can't detect when and what happens in a line like this after the call to RecordUndo: `settings.RecordUndo().milestones.Add(obj);`

I have another approach in mind, but that needs some bigger refactoring (same direction as @SWeinis idea about transactions).

PS: We had another use case for this in the Discord. Someone had a power outage IRL :/
So at least recurring autosaves would be a nice thing. Likes Factorio's, rolling around a ringbuffer...

**Comment by shihan42** [comment]
What do you people think about this: Autosave is supported in both flavors. Either every x minutes or after each meaningful operation (whatever that is).
Plus the option of no autsave whatsoever...

**Comment by have-fun-was-taken** [comment]
Sounds good to me.

**Comment by MartijnvAdrichem** [comment]
Hello, what is the current state of this PR? Since space age I started using YAFC but lost quite a bit of data because I kept forgetting to save it before shutting down my PC... 

Maybe I could help finishing the PR?

**Comment by shpaass** [comment]
Hi @MartijnvAdrichem,
You're free to pick it up. I think @shihan42 would not be against it.
Please keep in mind that the current code in the PR is not set in stone, as you can see from the discussion above.

**Comment by shihan42** [comment]
Yeah, absolutely. Take it, change it, do what you want 😀 

I'm very much out of free time ATM, so feel free to take over.

**Comment by MartijnvAdrichem** [comment]
@shihan42 Could you give a quick summary where you left it off with the last few commits?

**Comment by shihan42** [comment]
Nothing too fancy right there. I added an option to the configuration to enable/disable autosaves.
Then I began with async calls to enable async I/O. During that, I saw some problems with async and the (then) current GUI and rendering approach.
My ideas were to take a snapshot of the project (i.e., deep clone the object), start a background thread and save the cloned project to file.

Never made it that far, though. RL threw me off, so I stopped my development for yafc-ce. Haven't even tried Space Age, unfortunately 😞

**Comment by shpaass** [comment]
If the Thread option is too complex, what do you guys think about saving periodically on losing focus, or if the focus is not lost, then on user being AFK? 
I think this approach allows us to save as usual, because the in-Yafc lag would not matter there.

**Comment by MartijnvAdrichem** [comment]
I think I found quite an easy solution after reading your last post @shpaass 

Hopefully this patch file is a bit readable, I would like to know if you guys think this is the way forward.

How it works:
Window throws an event on focus lost 
Mainscreen listens to the event on focus lost and performs a Save() 
So everytime the user leaves the screen a save is performed, but because of the current save logic it only actually saves if anything was changed it won't spam too much. The only situation I could think of that this would not work is when you are working actively in YAFC and then the power goes off or something, but this does soften that problem.

If you guys see no issues with this solution I would like to add the following:
- Autosaving goes to a different filename (xxx-autosave-1.yafc), maybe with a rolling window of 5 saves. This keeps the "main" save open to manual saving when you are just toying around
- When opening a save look if there are any autosaves, open the last edited instead (maybe let the user choose if he wants to open the auto save or not)?

When I start working on that I will override the PR of @shihan42 :)

```c#

Subject: [PATCH] Autosave
---
Index: Yafc/Windows/PreferencesScreen.cs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/Yafc/Windows/PreferencesScreen.cs b/Yafc/Windows/PreferencesScreen.cs
--- a/Yafc/Windows/PreferencesScreen.cs	(revision ae70bc43d218f66ba1d0a4cc955296ec089548f6)
+++ b/Yafc/Windows/PreferencesScreen.cs	(date 1738252007768)
@@ -88,6 +88,12 @@
                     .OrderBy(x => x.locName)]
             };
         }
+        using (gui.EnterRow()) {
+            gui.BuildText("Autosave project:", topOffset: 0.25f);
+            if (gui.BuildCheckBox("enable", Project.current.preferences.autosaveEnabled, out bool newAutosave)) {
+                Project.current.preferences.autosaveEnabled = newAutosave; 
+            }
+        }
 
         technologyList.Build(gui);
         technologyList.RebuildContents();
Index: Yafc.UI/Core/Window.cs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/Yafc.UI/Core/Window.cs b/Yafc.UI/Core/Window.cs
--- a/Yafc.UI/Core/Window.cs	(revision ae70bc43d218f66ba1d0a4cc955296ec089548f6)
+++ b/Yafc.UI/Core/Window.cs	(date 1738831722385)
@@ -38,6 +38,8 @@
     public virtual bool preventQuit => false;
     internal Window(Padding padding) => rootGui = new ImGui(Build, padding);
 
+    public event OnFocusLost? onFocusLost;
+    
     internal void Create() {
         if (surface is null) {
             throw new InvalidOperationException($"surface must be set by a derived class before calling {nameof(Create)}.");
@@ -179,7 +181,8 @@
         }
     }
 
-    public virtual void FocusLost() { }
+    public virtual void FocusLost() => onFocusLost?.Invoke();
+    
     public virtual void Minimized() { }
 
     public void SetNextRepaint(long nextRepaintTime) {
@@ -263,3 +266,5 @@
         GC.SuppressFinalize(this);
     }
 }
+
+public delegate void OnFocusLost(); 
Index: Yafc/Windows/MainScreen.cs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/Yafc/Windows/MainScreen.cs b/Yafc/Windows/MainScreen.cs
--- a/Yafc/Windows/MainScreen.cs	(revision ae70bc43d218f66ba1d0a4cc955296ec089548f6)
+++ b/Yafc/Windows/MainScreen.cs	(date 1738830681941)
@@ -60,6 +60,8 @@
         Create("Yet Another Factorio Calculator CE v" + YafcLib.version.ToString(3), display, Preferences.Instance.initialMainScreenWidth,
             Preferences.Instance.initialMainScreenHeight, Preferences.Instance.maximizeMainScreen);
         SetProject(project);
+
+        this.onFocusLost += project.PerformAutoSave;
     }
 
     [MemberNotNull(nameof(project))]
Index: Yafc.Model/Model/Project.cs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/Yafc.Model/Model/Project.cs b/Yafc.Model/Model/Project.cs
--- a/Yafc.Model/Model/Project.cs	(revision ae70bc43d218f66ba1d0a4cc955296ec089548f6)
+++ b/Yafc.Model/Model/Project.cs	(date 1738831652288)
@@ -25,7 +25,7 @@
     public new UndoSystem undo => base.undo;
     private uint lastSavedVersion;
     public uint unsavedChangesCount => projectVersion - lastSavedVersion;
-
+    
     public Project() : base(new UndoSystem()) {
         settings = new ProjectSettings(this);
         preferences = new ProjectPreferences(this);
@@ -135,6 +135,13 @@
         SerializationMap<Project>.SerializeToJson(this, writer);
     }
 
+    public void PerformAutoSave() {
+        if (preferences.autosaveEnabled) {
+            Console.WriteLine("Auto-saving...");
+            Save(attachedFileName);
+        }
+    }
+
     public void RecalculateDisplayPages() {
         foreach (var page in displayPages) {
             FindPage(page)?.SetToRecalculate();
@@ -264,6 +271,8 @@
     public int maxMilestonesPerTooltipLine { get; set; } = 28;
     public bool showMilestoneOnInaccessible { get; set; } = true;
 
+    public bool autosaveEnabled { get; set; }
+    
     protected internal override void AfterDeserialize() {
         base.AfterDeserialize();
         defaultBelt ??= Database.allBelts.OrderBy(x => x.beltItemsPerSecond).FirstOrDefault();
Index: Yafc/Utils/Preferences.cs
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/Yafc/Utils/Preferences.cs b/Yafc/Utils/Preferences.cs
--- a/Yafc/Utils/Preferences.cs	(revision ae70bc43d218f66ba1d0a4cc955296ec089548f6)
+++ b/Yafc/Utils/Preferences.cs	(date 1738658614429)
@@ -11,9 +11,7 @@
     public static readonly Preferences Instance;
     public static readonly string appDataFolder;
     private static readonly string fileName;
-
-    public static readonly string autosaveFilename;
-
+    
     static Preferences() {
         appDataFolder = Environment.GetFolderPath(Environment.SpecialFolder.LocalApplicationData);
         if (!RuntimeInformation.IsOSPlatform(OSPlatform.OSX)) {
@@ -24,7 +22,6 @@
             _ = Directory.CreateDirectory(appDataFolder);
         }
 
-        autosaveFilename = Path.Combine(appDataFolder, "autosave2.yafc");
         fileName = Path.Combine(appDataFolder, "yafc2.config");
         if (File.Exists(fileName)) {
             try {


```

**Comment by shihan42** [comment]
The rolling window and the save on focus lost is a good idea, I think. I don't think there are many edge cases and if so, I'd assume they wouldn't matter.

So, go for it!

**Comment by shpaass** [comment]
> Console.WriteLine("Auto-saving...");

Please use a logger instead. There should be examples of the logger usage around the code. Let me know if you don't find them, I'll look for a concrete example.

**Comment by veger** [comment]
> How it works

Sounds reasonable to me, and should cover a majority of the issues as people usually (my assumption) switch to other windows (like Factorio) often and thus triggering an auto save.

> Autosaving goes to a different filename (xxx-autosave-1.yafc), maybe with a rolling window of 5 saves. This keeps the "main" save open to manual saving when you are just toying around

We could even have an option to delete the auto saves when a manual save is triggered. As the 'main' save is then up to speed with the auto saves. So it is not a backup system (anymore) but more a system to recover on (power) failures.

> When opening a save look if there are any autosaves, open the last edited instead (maybe let the user choose if he wants to open the auto save or not)?

With the cleaned auto saves, it is fairly easy to detect auto saves and warn the user about them when opening the 'main' save in a new session. Or even refusing to load until the user sorted it out.

Eventually letting them decide what to do, use a certain version (and clean up the rest).
Initially a warning could be enough, to keep this PR size acceptable, so the user can go to his saves and figure it out.

**Comment by DaleStan** [comment]
Closing this as it was superseded by #412
